Modules/SubModule_Import/xASL_imp_DCM2NII_Subject_ShuffleTheDynamics.m

```diff
@@ -130,40 +132,59 @@
         end
     end
     
-    %% FME (Hadamard) Check
-    if numel(nii_files)>=1
+    %% Hadamard Check
+    
+    bHadamard = false;
+    bHadamardFME = false;
+    
+    % Determine if we have a Hadamard sequence based on the parameters of the studyPar.json
+    if xASL_exist(x.dir.studyPar,'file')
+        studyPar = spm_jsonread(x.dir.studyPar);
+        if isfield(studyPar,'HadamardMatrixType')
+            bHadamard = true;
+        end
+    end
+    
+    
+    %% Check for specific Hadamard sequences and reorder the NIfTI volumes if necessary
+    if numel(nii_files)>=1 
         [resultPath, resultFile] = xASL_fileparts(nii_files{1});
```

Comment:
This is a very Hadamard-specific part, how does this work for non-Hadamard NIfTIs? You can have other ASL types that have `numel(nii_files)>=1` as well (in fact, you should always have at least one NIfTI...)

Reply:
I understand why this seems a bit misleading. So if there are the typical new Hadamard/TimeEncoded fields in the studyPar, we know that `bHadamard=true`/`bTimeEncoded=true` at that part already. Originally we detected that stuff from the FME `SeriesDescription` though. To make it a bit more readable, we could separate this part ...

```matlab
    %% Check for specific Hadamard sequences and reorder the NIfTI volumes if necessary
    if numel(nii_files)>=1 
        [resultPath, resultFile] = xASL_fileparts(nii_files{1});
        % Check if we have the corresponding JSON file
        if exist(fullfile(resultPath, [resultFile '.json']), 'file') 
        	% Load the JSON
            resultJSON = spm_jsonread(fullfile(resultPath, [resultFile '.json']));
            % Check if we have the SeriesDescription field
            if isfield(resultJSON,'SeriesDescription') || bHadamard
            	% Determine if we have the specific FME Hadamard sequence from Bremen
                bHadamardFME = ~isempty(regexp(resultJSON.SeriesDescription),'(Encoded_Images_Had)\d\d(_)\d\d(_TIs_)\d\d(_TEs)', 'once');
                % If the FME sequence was detected we can always set the general bHadamard to true as well
                if bHadamardFME
                    bHadamard = true;
                end
            end
...
```

... and this part ...


```matlab
...
            % Check if we the current sequence is a Hadamard or not
            if bHadamard
                % Check image
                if xASL_exist(nii_files{1},'file')
                    % Determine the number of time points within each NIfTI
                    imASL = xASL_io_Nifti2Im(nii_files{1});
                    numberTEs = length(resultJSON.EchoTime);
                    numberPLDs = int32(size(imASL,4)/numberTEs);
                    
                    % Reorder TEs and PLDs - first cycle TE afterwards PLD
                    vectorOldOrder = zeros(size(imASL,4),1);
                    for iPLD = 1:(double(numberPLDs))
                        vectorOldOrder((1:numberTEs)+(iPLD-1)*numberTEs) = (iPLD-1)+1:numberPLDs:size(imASL,4);
                    end
                    imASL(:,:,:,1:end) = imASL(:,:,:,vectorOldOrder);
                    xASL_io_SaveNifti(nii_files{1},nii_files{1},imASL);
                    % Repeat Echo Times
                    resultJSON.EchoTime = repmat(resultJSON.EchoTime,numberPLDs,1);
                    % Save the JSON with the updated echo times
                    spm_jsonwrite(fullfile(resultPath, [resultFile '.json']),resultJSON);
                else
                    % Fallback (something went wrong)
                    warning('Hadamard sequence with 1 PLD only');
                end
            end
        end
    end
```

... though.